### PR TITLE
tests: dependency packages installed during prepare-project

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The objective of this change is to provide an environment for all the
tests as similar as possible, so all the dependencies are installed
during the project setup.

The packages are not removed at the end of the project, so it will help
in case the machine needs to be used to debug.

Openvswitch test is moved to manual due to the dependencies install is
stuck trying to install openvswitch-switch package.

### CODE REVIEW

**Comment by mvo5** [comment]
Thanks for doing this! It looks good, I wonder however if dependencies.sh should instead simply be part of pkgdb.sh, this way all pkg releated things can be easily found in the same place.

File: tests/lib/dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,131 @@
+#!/bin/bash
+
+# shellcheck source=tests/lib/pkgdb.sh
+. "$TESTSLIB/pkgdb.sh"
+
+export DEPENDENCY_PACKAGES=
+export DISTRO_BUILD_DEPS=
+
+add_pkgs(){
+  DEPENDENCY_PACKAGES="$DEPENDENCY_PACKAGES $@"
```
Comment (niemeyer) [inline tests/lib/dependencies.sh]: We're trying to standardize on four spaces per indent on shell and yaml code. I've seen some other files which are already all over the place, with mixed 2, 3 and 4 spaces on the same file. Let's please fix that as we drive by these files (no need to rush it), and also try not to introduce new inconsistencies meanwhile.

File: tests/lib/dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,131 @@
+#!/bin/bash
```
Comment (niemeyer) [inline tests/lib/dependencies.sh]: Agree with @mvo5 that pkgdb.sh seems like a reasonable place for this logic.

File: tests/lib/dependencies.sh [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+export DEPENDENCY_PACKAGES=
+export DISTRO_BUILD_DEPS=
+
+add_pkgs(){
+  DEPENDENCY_PACKAGES="$DEPENDENCY_PACKAGES $@"
+}
+
+get_apt_dependencies_generic(){
+  add_pkgs autoconf
```
Comment (niemeyer) [inline tests/lib/dependencies.sh]: Communicating across shell functions via sticky environment variables is pretty hard to follow. It's a backdoor without a clear flow between the involved parties. Instead, this can simply `echo autoconf` here, and the other side can use the standard output as usual for its needs.

File: tests/lib/dependencies.sh [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+get_dependency_fedora_packages(){
+  echo "Fedora dependencies not ready yet"
+}
+
+get_dependency_opensuse_packages(){
+  echo "Opensuse dependencies not ready yet"
+}
+
+get_test_dependencies(){
```
Comment (niemeyer) [inline tests/lib/dependencies.sh]: Is there an advantage in splitting down test and build?  Won't we end up with both of these installed anyway?

File: tests/lib/dependencies.sh [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+          add_pkgs pollinate
+          add_pkgs printer-driver-cups-pdf
+          ;;
+      debian-*)
+          add_pkgs printer-driver-cups-pdf
+          ;;
+  esac 
+}
+
+get_apt_dependencies_core(){
```
Comment (niemeyer) [inline tests/lib/dependencies.sh]: I suggest s/get_/pkg_/ on all of these functions. This preserves them with the same length and adds a bit of context to it (e.g. "test dependencies" alone could mean many things).

File: tests/lib/dependencies.sh (hunk 2 above)
Comment (sergiocazzolato) [inline tests/lib/dependencies.sh]: sure, make sense to me too

File: tests/lib/dependencies.sh (hunk 4 above)
Comment (sergiocazzolato) [inline tests/lib/dependencies.sh]: yes, I already put all of them together

**Comment by chipaca** [comment]
Looks like the debian failures are related to your changes.

**Comment by sergiocazzolato** [comment]
@niemeyer when you have time, please take a look to this one, all the comments were addressed.

File: tests/lib/pkgdb.sh [hunk 6]
```diff
@@ -305,19 +305,124 @@ distro_install_build_snapd(){
...
-    fedora-*)
-        DISTRO_BUILD_DEPS=(mock git expect curl golang rpm-build redhat-lsb-core)
-        ;;
-    opensuse-*)
-        DISTRO_BUILD_DEPS=(osc git expect curl golang-packaging lsb-release netcat-openbsd jq rng-tools)
-        ;;
-    *)
-        ;;
-esac
+pkg_apt_dependencies_generic(){
```
Comment (fgimenez) [inline tests/lib/pkgdb.sh]: I find this naming a bit inconsistent with what we have for fedora and opensuse where there's no reference to rpm, maybe `pkg_dependency_ubuntu_{generic,classic}` would be better.

Comment (sergiocazzolato) [inline tests/lib/pkgdb.sh]: sure

File: tests/lib/pkgdb.sh [hunk 7]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
-        ;;
-    opensuse-*)
-        DISTRO_BUILD_DEPS=(osc git expect curl golang-packaging lsb-release netcat-openbsd jq rng-tools)
-        ;;
-    *)
-        ;;
-esac
+
+pkg_dependency_ubuntu_generic(){
+    echo autoconf
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: On all these cases, this would be nicer:
```
echo "
    foo
    bar
    baz
    "
```

File: tests/lib/pkgdb.sh [hunk 8]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
+    echo curl
+    echo expect
+    echo git
+    echo golang
+    echo mock
+    echo redhat-lsb-core
+    echo rpm-build
+}
+
+pkg_dependency_opensuse(){
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: s/dependency/dependencies/ on all cases above would read slightly better and also align with the cases below.

File: tests/lib/pkgdb.sh [hunk 9]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
+        opensuse-*)
+            pkg_dependency_opensuse
+            ;;
+        *)
+            ;;
+    esac  
+}
+
+install_dependencies(){
+    pkgs=$(pkg_dependencies | tr "\n" " ")
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: The tr should not be necessary here.

File: tests/lib/pkgdb.sh [hunk 10]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
+
+    # ensure systemd is up-to-date, if there is a mismatch libudev-dev
+    # will fail to install because the poor apt resolver does not get it
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: As such, this should be inside distro_install_package and should be conditional on libudev-dev being part of the requested set.

File: tests/lib/pkgdb.sh [hunk 11]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
+    case "$SPREAD_SYSTEM" in
+        debian-*|ubuntu-*)
+            apt-get install -y --only-upgrade systemd
+    esac
+
+    echo "Installing the following packages: $pkgs"
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: The logging should also be inside the install function.

File: tests/lib/pkgdb.sh [hunk 12]
```diff
@@ -305,19 +305,125 @@ distro_install_build_snapd(){
...
+            ;;
+        opensuse-*)
+            pkg_dependency_opensuse
+            ;;
+        *)
+            ;;
+    esac  
+}
+
+install_dependencies(){
```
Comment (niemeyer) [inline tests/lib/pkgdb.sh]: install_pkg_dependencies perhaps?

**Comment by sergiocazzolato** [comment]
@niemeyer fixes requested already pushed
